Sean's 9/20 work

seaan opened the pull request:
implemented a concerning amount of things...

-Moved all driver .c and .h files to their own folders for easier management
EXAMPLE: ADC_init.c and ADC_init.h are now both in the Drivers/ADC/ directory.

-Created getAltitude and getVelocity .c and .h files, as well as methods.

-Created flightState, altitude, pressure, temperature, initialAltitude variables.

-Cleaned main.c, added #include for new and improved drivers.

-Added switch case for flightState to main.c

-Created flightState class; this houses flightState 0-3 methods, created a shell for each along with a comment or rough code for what each needs to do. Also began writing the code for the descent flight state (2).

-Probably also other minor things that I forgot.

-Wew lad

seaan committed:
Sean's 9/20 work

implemented a concerning amount of things...

-Moved all driver .c and .h files to their own folders for easier
management
EXAMPLE: ADC_init.c and ADC_init.h are now both in the Drivers/ADC
folder

-Created getAltitude and getVelocity .c and .h files, as well as
methods.

-Created flightState, altitude, pressure, temperature, initialAltitude
variables.

-Cleaned main.c, added #include for new and improved drivers.

-Added switch case for flightState to main.c

-Created flightState class; this houses flightState 0-3 methods, created
a shell for each along with a comment or rough code for what each needs
to do. Also began writing the code for the descent flight state (2).

-Probably also other minor things that I forgot.

seaan committed:
Fixed a lot of errors, added descent control .c and .h

Can't seem to fix implicit declaration of FSZero, etc. in main.c :100:

seaan committed:
Add PseudoCode + Software Flowchart files to be modified

EinaamAlim commented:
Whew, lad, indeed. Okay, so reading through this, it looks like we're on our way.

Is there anything in particular that we NEED to revise and/or implement?

seaan committed:
Holy crap 9/21 revised after PDR, still missing a lot of stuff tho lol

-Revised flight state case switch to a set of recursive statements for
each flight state. In flightState.c, I changed the while loops to simple
if statements asking if we were at the altitude required for the flight
state. If we are, then you run the function again (RECURSION YAY)

-Revised flight state 2 method into a waaay more simple function. The
only reason we need to check our velocity is to see if we need to deploy
the chutes.. So, it's a nice little if statement asking if we're above
the death velocity, and if we are, deploy chutes. This allows us to also
check the velocity, and probably write the velocity to eeprom within the
statement but idk we can figure it out later.

-getVelocity().. Ahh, I longed for the times it was a simple delta Alt
over delta T. Well, instead, taking Daniel's advice, I wrote a
getVelocity function that uses exponential smoothing to reduce the noise
from all that derivative taking.

See Exponential Smoothing here PLEASE:
https://en.wikipedia.org/wiki/Exponential_smoothing

--More on the exponential smoothing.. So what I did was create two array
tables, one for the rough velocity measurements and one for the smoothed
measurements. We can then use the smoothed measurements to take an
average for velocity over 250ms which I think is pretty decent.

--- Now onto how I actually did it.. First we iterate through each
element of our rough table, assigning a velocity to each one. We find
the velocity by taking the difference between altitude at the beginning
of our iteration and at the end of the previous iteration.. (For the
first iteration, we're gonna throw that one out), We then divide this
change in altitude by the change in time, which we convieniently always
know because I put in a delay of 10ms. This means that we can take 25
samples over 250ms, which is a 4Hz sampling rate, exactly what we wanted
(I think lol). HOORAY we have a table for velocity that is ridiculously
noisy! Luckily, we can smooth it. We do this by iterating through each
element in our smoothed table, and assigning it equal to the value that
our exponentialSmoothing function spits out after we give it the current
position's rough table element as well as the previous smoothed table
element. This lets us preform the computation that wikipedia gave me
pretty easily (I think lol), and we can then just add up all the values
of our new table and average it. WEW LAD

-Conclusion: If this works, I am a literal programming god
No seriously, this is the most complex computation that I have ever
done.
Like 50x the complexity of the next most compex computation.
It took me only 2 hours to write this I have no idea how.
It should have taken me a lot more, and I should have actually planned
and designed it.
So again, if this works, holy crap I'm an actual coding god.

Once again,
-Wew lad

seaan committed:
correction

seaan committed:
attempted to write getTemperature(), can't get the right values D:

seaan committed:
Fixed getTemperature(), started on getAltitude()

seaan committed:
alt adjustment
